File: Code/bitcmp/bitcmp.py
import os
import sys
import threading
import concurrent.futures
from time import sleep

NUM_THREADS = 4
CHUNK_SIZE = 1024 * 1024  # 1MB per read


def count_bits(byte):
    return bin(byte).count('1')


def compare_chunks(data1, data2, progress, lock, bit_diff_total):
    local_diff = 0
    for b1, b2 in zip(data1, data2):
        local_diff += count_bits(b1 ^ b2)
    with lock:
        bit_diff_total[0] += local_diff
        progress[0] += len(data1)


def print_progress(progress, total_size, stop_flag):
    while not stop_flag[0]:
        percent = progress[0] / total_size * 100
        print(f"\rProgress: {percent:.1f}% ({progress[0]} / {total_size} bytes)", end='', flush=True)
        sleep(0.2)
    print(f"\rProgress: 100.0% ({total_size} / {total_size} bytes)")


def main(file1, file2):
    if not os.path.exists(file1) or not os.path.exists(file2):
        print("Error: One or both files not found")
        return

    size1 = os.path.getsize(file1)
    size2 = os.path.getsize(file2)
    if size1 != size2:
        print("Error: Files are not the same size")
        return

    progress = [0]
    bit_diff_total = [0]
    lock = threading.Lock()
    stop_flag = [False]

    t = threading.Thread(target=print_progress, args=(progress, size1, stop_flag))
    t.start()

    with open(file1, 'rb') as f1, open(file2, 'rb') as f2:
        with concurrent.futures.ThreadPoolExecutor(max_workers=NUM_THREADS) as executor:
            futures = []
            while True:
                chunk1 = f1.read(CHUNK_SIZE)
                chunk2 = f2.read(CHUNK_SIZE)
                if not chunk1:
                    break
                futures.append(executor.submit(compare_chunks, chunk1, chunk2, progress, lock, bit_diff_total))

            concurrent.futures.wait(futures)

    stop_flag[0] = True
    t.join()

    print(f"\nDiffering bits: {bit_diff_total[0]}")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python bitcmp.py file1 file2")
        sys.exit(1)
    main(sys.argv[1], sys.argv[2])

File: Code/bitcmp/bitcmp.c
/******************************************************************************
 *
 * Name: bitcmp.c - Description
 * Created on 2025/06/06
 * Copyright (C) 2022 - 2025, wyh.
 *
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <pthread.h>
#include <sys/stat.h>
#include <unistd.h>
#include <time.h>

#define NUM_THREADS 4
#define PROGRESS_INTERVAL_MS 200  // 进度显示刷新时间，毫秒

typedef struct {
	uint8_t *buf1;
	uint8_t *buf2;
	size_t start;
	size_t end;
	size_t bit_diff;
} ThreadArg;

volatile size_t total_processed = 0;
size_t file_size = 0;
volatile int done = 0;

// 统计一个字节中置1的比特数
int count_bits(uint8_t byte)
{
	int count = 0;
	while (byte) {
		count++;
		byte &= (byte - 1);
	}
	return count;
}

// 子线程：执行比特比较
void *compare_bits(void *arg)
{
	ThreadArg *targ = (ThreadArg *)arg;
	size_t diff = 0;

	for (size_t i = targ->start; i < targ->end; i++) {
		uint8_t xor = targ->buf1[i] ^ targ->buf2[i];
		diff += count_bits(xor);
		__sync_fetch_and_add(&total_processed, 1); // 原子增加处理计数
	}

	targ->bit_diff = diff;
	return NULL;
}

// 进度监视线程
void *progress_monitor(void *arg)
{
	(void)arg;
	while (!done) {
		usleep(PROGRESS_INTERVAL_MS * 1000); // 毫秒
		size_t processed = total_processed;
		double percent = (double)processed / file_size * 100.0;
		printf("\rProgress: %.1f%% (%zu / %zu bytes)   ", percent,
		       processed, file_size);
		fflush(stdout);
	}
	// 最后打印100%
	printf("\rProgress: 100.0%% (%zu / %zu bytes)\n", file_size, file_size);
	return NULL;
}

int main(int argc, char *argv[])
{
	if (argc != 3) {
		fprintf(stderr, "Usage: %s <file1> <file2>\n", argv[0]);
		return EXIT_FAILURE;
	}

	FILE *f1 = fopen(argv[1], "rb");
	FILE *f2 = fopen(argv[2], "rb");
	if (!f1 || !f2) {
		perror("Error opening files");
		return EXIT_FAILURE;
	}

	struct stat st1, st2;
	fstat(fileno(f1), &st1);
	fstat(fileno(f2), &st2);

	if (st1.st_size != st2.st_size) {
		fprintf(stderr, "Error: Files are not the same size\n");
		return EXIT_FAILURE;
	}

	file_size = st1.st_size;
	uint8_t *buf1 = malloc(file_size);
	uint8_t *buf2 = malloc(file_size);
	if (!buf1 || !buf2) {
		perror("Memory allocation failed");
		return EXIT_FAILURE;
	}

	fread(buf1, 1, file_size, f1);
	fread(buf2, 1, file_size, f2);
	fclose(f1);
	fclose(f2);

	// 启动进度线程
	pthread_t progress_thread;
	pthread_create(&progress_thread, NULL, progress_monitor, NULL);

	// 启动比较线程
	pthread_t threads[NUM_THREADS];
	ThreadArg args[NUM_THREADS];
	size_t chunk = file_size / NUM_THREADS;

	for (int i = 0; i < NUM_THREADS; i++) {
		args[i].buf1 = buf1;
		args[i].buf2 = buf2;
		args[i].start = i * chunk;
		args[i].end = (i == NUM_THREADS - 1) ?
			              file_size :
			              (i + 1) * chunk;
		args[i].bit_diff = 0;
		pthread_create(&threads[i], NULL, compare_bits, &args[i]);
	}

	size_t total_diff = 0;
	for (int i = 0; i < NUM_THREADS; i++) {
		pthread_join(threads[i], NULL);
		total_diff += args[i].bit_diff;
	}

	done = 1;
	pthread_join(progress_thread, NULL);

	printf("Differing bits: %zu\n", total_diff);

	free(buf1);
	free(buf2);
	return EXIT_SUCCESS;
}
